### DISCUSSION TITLE
Running multiple animations on a single model + rig

### DISCUSSION THREAD

**Discussion by ninjz** [sceneview/sceneview#326] [Q&A]
Is it possible to apply different animations to a single model + rig? I've played around but I can't seem to get this to work. Currently I need to include the full mesh, rig and animation for each animation. I also do not want to have all the animations within one file with the mesh because that file itself would be huge.

Does this library support this functionality?

**Comment by ThomasGorisse** [comment]
I don't totally get what you want to achieve.
Could you please share some code you'd like to implement?

**Reply by ninjz** [reply]
Thanks for getting back to me on this @ThomasGorisse. I have a sample snippet of what I have on my iOS app using SceneKit, of which I'm trying to achieve with this library:

```swift
import Foundation
import SceneKit

class AnimationPlayerView: SCNView {
    // Animation controlling node
    var animationNode: SCNNode?

    override public init(frame: CGRect, options: [String: Any]? = nil) {
        super.init(frame: frame, options: options)
        do {
            let url = Bundle.module.url(forResource: "art.scnassets/Male_Character", withExtension: "usdz")
            
            self.scene = try SCNScene(url: url!, options: nil)
        } catch {
            print("Fatal Error \(error)")
        }
        
        self.delegate = self

        // Animation skeleton - node where animatable rig is attached
        animationNode = self.scene!.rootNode.childNode(withName: "Skeleton", recursively: true)
    }
    
    public func playAnimation(id: String) {
        // Load animation from file
        let url = Bundle.module.url(forResource: "art.scnassets/\(id)", withExtension: "usdz")
        do {
            let scene = try SCNScene(url: url!, options: nil)
            var animation: SCNAnimation? = nil
            
            scene.rootNode.enumerateChildNodes { child, _ in
                if !child.animationKeys.isEmpty {
                    animation = child.animationPlayer(forKey: child.animationKeys[0])?.animation
                }
            }
            
            let player = SCNAnimationPlayer(animation: animation!)
            
            // Attach animation player to main Skeleton so the animation will play on the target rig
            animationNode?.addAnimationPlayer(player.copy() as! SCNAnimationPlayer, forKey: id)
        } catch {
            print("Error loading animation: \(error)")
        }

        // Play animation
        let player = animationNode?.animationPlayer(forKey: id)
    
        // Play animation
        player?.play()
    }
}
```

You can see here that on init, the scene is being set up with a `usdz` scene which contains a complete character with mesh, rig and textures. 

When calling `playAnimation()` a `usdz` scene is loaded which contains only animation data and no mesh, rig, and textures. Here
the animation data is loaded and applied on the character loaded on init.

Is this more clear?

**Reply by ninjz** [reply]
Any thoughts on this?

**Reply by danielok1993** [reply]
@ninjz have you managed to get this to work? I am also looking for a solution to play animations from different files onto the main file with the mesh

**Reply by ninjz** [reply]
@danielok1993 Ended up going down to Filament directly but we had to make some modifications to get it working with Filament. We submitted a PR but it never got merged: https://github.com/google/filament/pull/7919

**Comment by ThomasGorisse** [comment]
In v4.0.0-rc.1, GLB files can contain multiple animations. Access them by index via the Filament `Animator`:

```kotlin
val model = rememberModelInstance(modelLoader, "models/character.glb")
if (model != null) {
    val animator = model.animator
    val animCount = animator?.animationCount ?: 0
    // Play a specific animation by index
    animator?.playAnimation(0) // idle
    // Or switch to another
    animator?.playAnimation(1) // walk
}
```

For blending animations (e.g. walk + wave simultaneously), Filament supports cross-fading between animations via `applyAnimation` with blend weights. Export all animations in a single GLB from your 3D tool for best results.
